plugins/map_matching.hpp

```diff
@@ -0,0 +1,314 @@
+        }
+
+        subtrace.values["indices"] = osrm::json::make_array(sub.indices);
+
+
+        osrm::json::Array points;
+        for (const auto& node : sub.nodes)
+        {
+            points.values.emplace_back(osrm::json::make_array(node.location.lat / COORDINATE_PRECISION,
+                                                              node.location.lon / COORDINATE_PRECISION));
+        }
+        subtrace.values["matched_points"] = points;
+
+        return subtrace;
+    }
+
+    int HandleRequest(const RouteParameters &route_parameters, osrm::json::Object &json_result) final
+    {
+        // check number of parameters
+        if (!check_all_coordinates(route_parameters.coordinates))
+        {
+            return 400;
+        }
+
+        std::vector<double> sub_trace_lengths;
+        Matching::CandidateLists candidates_lists;
+        const auto& input_coords = route_parameters.coordinates;
+        const auto& input_timestamps = route_parameters.timestamps;
+        if (input_timestamps.size() > 0 && input_coords.size() != input_timestamps.size())
+        {
+            return 400;
+        }
+
+        // enforce maximum number of locations for performance reasons
+        if (max_locations_map_matching > 0 && static_cast<int>(input_coords.size()) > max_locations_map_matching)
+        {
+            return 400;
+        }
+
+        bool found_candidates = getCandiates(input_coords, sub_trace_lengths, candidates_lists);
+        if (!found_candidates)
+        {
+            return 400;
+        }
+
+        // setup logging if enabled
+        if (osrm::json::Logger::get())
+            osrm::json::Logger::get()->initialize("matching");
+
+        // call the actual map matching
+        Matching::SubMatchingList sub_matchings;
+        search_engine_ptr->map_matching(candidates_lists,
+                                        input_coords,
+                                        input_timestamps,
+                                        route_parameters.matching_beta,
+                                        route_parameters.gps_precision,
+                                        sub_matchings);
+
+        if (1 > sub_matchings.size())
+        {
+            return 400;
+        }
+
+        osrm::json::Array matchings;
+        for (auto& sub : sub_matchings)
+        {
+            // classify result
+            if (route_parameters.classify)
+            {
+                double trace_length = sub_trace_lengths[sub.indices.back()] - sub_trace_lengths[sub.indices.front()];
+                TraceClassification classification = classify(trace_length,
+                                                              sub.length,
+                                                              (sub.indices.back() - sub.indices.front() + 1) - sub.nodes.size());
+                if (classification.first == ClassifierT::ClassLabel::POSITIVE)
```

Comment:
Would it make sense to switch the two compared items? That way there wouldn't be an lvalue on the left-hand side and accidental assignments would (or could) not happen.